### PR TITLE
Rewrite compileStreaming/instantiateStreaming to be more rigorous

### LINKED ISSUE #1081: Unclear how to resolve/reject promises for response based compile API
Right now, the spec says that the API will run an asynchronous task that eventually resolves the promise. It's a bit unclear to me what an asynchronous task should be, it would be great to explicitly reference an event loop model to explain what this is.

Furthermore, in HTML's event loop model, only regular tasks enter a microtask checkpoint. Without doing this, resolved promise will be left hanging until something else spins the event loop again.

I'd expect that the spec here uses similar terms as e.g. fetch, i.e, when a response promise is passed in and the method doesn't immediately abort, it will finish the compilation asynchronously and when the module is ready create a task on the main event loop and that task resolves the promise.

### PR DESCRIPTION
**Pull request by domenic**
Fixes #1081 by being very clear about the interaction with the event loop, i.e. which portions of the tasks happen in parallel and how they get back to the main thread, where objects can actually be created. In general this also makes the specification algorithmic, links it to appropriate definitions, and makes various failure cases (such as being passed a rejected promise, or attempting to read from a locked/disturbed body) clearer.

/cc @jeisinger. @annevk may also enjoy this when he's back from vacation.

### CODE REVIEW

**Comment by jfbastien** [comment]
Is this intended to be an editorial clarification of something we'd agreed on, but which wasn't specified tightly enough?

**Comment by domenic** [comment]
Pretty much... One could argue it's a normative change because previously it wasn't really specified how some of this stuff worked (e.g. the previous spec kind of implied by omission that locked body streams would just be magically read by the UA, instead of throwing as this PR specifies). But IMO there should be nothing controversial in here.

**Comment by jfbastien** [comment]
OK, I think this can be reviewed as an editorial change since (I think?) only Chrome implements this API and it's clarifying what was ill-specified. If others discover issues after it's merged, we should be willing to revisit specific issues.

Does that sound right to you?

**Comment by domenic** [comment]
That does sound reasonable. I'd appreciate a check on the semantics from @jeisinger and probably @mtrofin though, in addition to any editorial review.

**Comment by jeisinger** [comment]
the changes lgtm

File: Web.md [hunk 1]
```diff
@@ -90,48 +74,133 @@ dictionary WebAssemblyInstantiatedSource {
...
+- _successSteps_ and _failureSteps_ will occur in tasks posted back to the main event loop; they
+  accept the Realm-agnostic values returned from _processingSteps_, and must transform them into
+  JavaScript objects that can be used to appropriately fulfill or reject the returned promise.
+
+Given these values, to process a potential WebAssembly response:
+
+1. Let _returnValue_ be [a new
+   promise](https://www.w3.org/2001/tag/doc/promises-guide#a-new-promise).
+1. Let _sourceAsPromise_ be [a promise resolved
+   with](https://www.w3.org/2001/tag/doc/promises-guide#a-promise-resolved-with) _source_.
```
Comment (equalsJeffH) [inline Web.md]: given the input description above on new line 129, should "source" actually be "argument" in these variable names?

Comment (domenic) [inline Web.md]: Great catch, thanks; fixing and force-pushing.

**Comment by domenic** [comment]
Let's merge, as this has @jeisinger's signoff and @mtrofin seems to be busy. He can follow up later if he has time to review...

**Pull request merged**